apps/web/src/pages/admin/users/index.tsx

```diff
@@ -0,0 +1,215 @@
+        id,
+        label: t(`users.roles.${id}`),
+      })),
+    []
+  )
+
+  const qp = getUsersFilterQuery({
+    page,
+    sortBy,
+    sortDirection,
+    pageSize: USERS_PER_PAGE,
+    search: searchDebounced,
+  })
+  const { data, isValidating } = useAuthApi<UserAccounts>(
+    `${urls.api.v1.admin.getUsers}?${qp}`
+  )
+  const users = data?.users || []
+  const total = data?.total || 0
+
+  const renderClaims = useCallback(
+    ({ item }: RendererProps) =>
+      rolesArray.map((role) => (
+        <CheckRole
+          key={`${item.externalId}-${role.id}`}
+          user={item}
+          role={role}
+        />
+      )),
+    []
+  )
+
+  const submitSearch = useCallback(
+    debounce((value) => {
+      setSearchDebounced(value)
+    }, 200),
+    []
+  )
+
+  const handleSearchChange = useCallback((value) => {
+    setSearch(value)
+    submitSearch(value)
+  }, [])
+
+  const columns: ColumnDefinitionType<UserAccount>[] = [
+    {
+      key: 'username',
+      name: t('users.table.Username'),
+      sortable: true,
+    },
+    {
+      key: 'email',
+      name: t('users.table.Email'),
+      sortable: true,
+    },
+    {
+      key: 'createdAt',
+      name: t('users.table.Created'),
+      renderer: ({ value }) =>
+        value ? new Date(value).toLocaleDateString(lang) : null,
+      sortable: true,
+    },
+    {
+      key: 'claims',
+      name: t('users.table.Role'),
+      renderer: renderClaims,
+    },
+  ]
+
+  const searchBox = (
+    <div className={'m-5'}>
+      <TextInput
+        handleChange={handleSearchChange}
+        placeholder={'Search by email or username'}
+        value={search}
+      />
+    </div>
+  )
+
+  const footer = (
+    <>
+      <div>
+        <Pagination
+          currentPage={page}
+          total={total}
+          pageSize={USERS_PER_PAGE}
+          setPage={setPage}
+        />
+      </div>
+      {total > 0 && <div>{total} records found</div>}
+    </>
+  )
+
+  return (
+    <AdminLayout pageTitle={t('common:pageTitles.Users')}>
+      <Breadcrumbs
+        breadcrumbs={[
+          { label: 'Admin', href: urls.admin.index },
+          { label: 'Users' },
+        ]}
+      />
+      <Panel
+        fullWidth
+        title={t('common:pageTitles.Users')}
+        contentRight={
+          isValidating && <RefreshIcon className="w-5 h-5 animate-spin" />
+        }
+        footer={footer}
+      >
+        {searchBox}
+        <div className="overflow-x-auto">
+          <Table<UserAccount>
+            columns={columns}
+            data={users}
+            onHeaderClick={handleTableHeaderClick}
+            sortBy={sortBy}
+            sortDirection={sortDirection as any}
+          />
+        </div>
+      </Panel>
+    </AdminLayout>
+  )
+}
+
+export const getServerSideProps: GetServerSideProps = async (context) => {
+  // Check if the user is admin (check again on render, to prevent caching of claims)
+  const user = await isAuthenticatedUserAdmin(context)
```

Comment:
Since this is an admin page though, shouldn't we kick a user out ASAP? Can also explore Next's new middleware feature: https://nextjs.org/docs/middleware

Reply:
When we tried to do this before it was way too aggressive (often kicked me out even though I _was_ logged in as an admin). I'm not sure why this happened but based on the behavior, my guess would be something to do with refresh tokens.

Reply:
Ideally, we'd be able to do this on the backend but if it's actually not possible (ie: firebase needs to look at cookies or something), I don't really see a problem with a client-side redirect.

Reply:
Ok, so https://github.com/rocketinsights/algomart-spin/pull/119/commits/6db3b39ca6e0f4d3d52a46c0125cc76e29d48e2e#diff-cefad03fa3af1123a70b2361f78a0e0acc695612b5902571ec5170c60f17e8daL199-L215 removes `getServerSideProps`. The page is wrapped in [AdminLayout](https://github.com/rocketinsights/algomart-spin/blob/6db3b39ca6e0f4d3d52a46c0125cc76e29d48e2e/apps/web/src/layouts/admin-layout.tsx#L21) which uses `useAdmin({ redirectIfNotAdmin: true })`. I also tried adding the `useAdmin()` to the AdminUsersPage component; in either case, a logged-out user can see the page (but not the users in the table). I'm new-ish to Next, so I'll fiddle a bit - just dropping here in case of obvious solution